v2/pkg/templates/compile.go

```diff
@@ -151,14 +101,98 @@ func Parse(filePath string, preprocessor Preprocessor, options protocols.Execute
 }
 
 // parseSelfContainedRequests parses the self contained template requests.
-func (t *Template) parseSelfContainedRequests() {
-	if !t.SelfContained {
+func (template *Template) parseSelfContainedRequests() {
+	if !template.SelfContained {
 		return
 	}
-	for _, request := range t.RequestsHTTP {
+	for _, request := range template.RequestsHTTP {
 		request.SelfContained = true
 	}
-	for _, request := range t.RequestsNetwork {
+	for _, request := range template.RequestsNetwork {
 		request.SelfContained = true
 	}
 }
+
+// Requests returns the total request count for the template
+func (template *Template) Requests() int {
+	return len(template.RequestsDNS) +
+		len(template.RequestsHTTP) +
+		len(template.RequestsFile) +
+		len(template.RequestsNetwork) +
+		len(template.RequestsHeadless) +
+		len(template.Workflows)
+}
+
+// compileProtocolRequests compiles all the protocol requests for the template
+func (template *Template) compileProtocolRequests(options protocols.ExecuterOptions) error {
+	templateRequests := template.Requests()
+
+	if templateRequests == 0 {
+		return fmt.Errorf("no requests defined for %s", template.ID)
+	}
+
+	if options.Options.OfflineHTTP {
+		template.compileOfflineHTTPRequest(options)
+		return nil
+	}
+
+	var requests []protocols.Request
+	switch {
+	case len(template.RequestsDNS) > 0:
+		requests = template.convertRequestToProtocolsRequest(template.RequestsDNS)
+
+	case len(template.RequestsFile) > 0:
+		requests = template.convertRequestToProtocolsRequest(template.RequestsFile)
+
+	case len(template.RequestsNetwork) > 0:
+		requests = template.convertRequestToProtocolsRequest(template.RequestsNetwork)
+
+	case len(template.RequestsHTTP) > 0:
+		requests = template.convertRequestToProtocolsRequest(template.RequestsHTTP)
+
+	case len(template.RequestsHeadless) > 0 && options.Options.Headless:
+		requests = template.convertRequestToProtocolsRequest(template.RequestsHeadless)
+	}
+	template.Executer = executer.NewExecuter(requests, &options)
+	return nil
+}
+
+// convertRequestToProtocolsRequest is a convenience wrapper to convert
+// arbitrary interfaces which are slices of requests from the template to a
+// slice of protocols.Request interface items.
+func (template *Template) convertRequestToProtocolsRequest(requests interface{}) []protocols.Request {
```

Comment:
This is not safe, nor should be needed. Every concrete type should fully implement the `protocols.Request` interface first (which is not the case currently)